neuralpp/experiments/experimental_inference/exact_belief_propagation.py

```diff
@@ -31,3 +33,64 @@ def variables_summed_out_at(self, node, all_variables_in_product_at_node):
 class ExactBeliefPropagation(BeliefPropagation):
     def __init__(self, factors, query):
         super().__init__(LazyFactorSpanningTree(FactorGraph(factors), query))
+
+
+class IncrementalAnytimeBeliefPropagation(PartialTreeComputation):
+
+    def __init__(self, partial_tree, full_tree, approximation_fn, expansion_fn):
+        super().__init__(partial_tree, full_tree)
+        self.approximation = approximation_fn
+        self.expansion = ExpansionValueComputation(partial_tree, full_tree, expansion_fn)
+        self.compute_result_dict(partial_tree.root)
+
+    @staticmethod
+    def from_factors(factors, query, approximation_fn, expansion_fn):
+        """
+        Simple utility to start AnytimeBeliefPropagation on a partial tree containing just the root.
+        """
+        full_tree = LazyFactorSpanningTree(FactorGraph(factors), query)
+        partial_tree = PartialFactorSpanningTree(FactorGraph(factors), query)
+        return IncrementalAnytimeBeliefPropagation(partial_tree, full_tree, approximation_fn, expansion_fn)
+
+    def compute_result_dict(self, node):
+        self[node] = None
+        self.message_from(node)
+
+    def update_value(self, target_node, changed_child, child_value):
```

Comment:
Yes, both methods are overrides from the TreeComputation. For belief propagation they're identical because it doesn't matter which child is changed or what the child's value is. On the other hand, the expansion computation cares about a new child_value and whether it's higher than the ancestor's existing value.

`changed_child` isn't actually used by either computation - I added it anticipating that some tree computations might want it, but could just take that out.